tiledb/sm/array/array.cc

```diff
@@ -1560,5 +1560,157 @@ void Array::set_serialized_array_open() {
   is_open_ = true;
 }
 
+void Array::get_max_tile_size(
+    tiledb_fragment_max_tile_sizes_t* max_tile_sizes) {
+  assert(max_tile_sizes);
+
+  if (remote_) {
+    auto rest_client = storage_manager_->rest_client();
+    if (rest_client == nullptr) {
+      throw StatusException(
+          Status_Error("Failed to obtain maximum tile size; remote array with "
+                       "no REST client."));
+    }
+
+    throw_if_not_ok(rest_client->get_array_maximum_tile_size_from_rest(
+        array_uri_, &max_tile_sizes->max_in_memory_tile_size));
+    return;
+  }
+
+  // Check that array is open
+  if (!is_open()) {
+    throw StatusException(Status_ArrayError(
+        "[Array::array_get_fragment_tile_max_size] Array is closed"));
+  }
+
+  if (query_type_ != QueryType::READ) {
+    throw StatusException(Status_ArrayError(
+        "[Array::array_get_fragment_tile_max_size] needs QueryType::READ"));
+  }
+
+  memset(max_tile_sizes, 0, sizeof(*max_tile_sizes));
+
+  EncryptionKey enc_key(config_);
+  uint64_t timestamp_start = 0;
+  uint64_t timestamp_end = UINT64_MAX;
+
+  // Create an ArrayDirectory object and then load the array schemas
+  // and fragment metadata.
+  auto array_dir =
+      ArrayDirectory(resources_, array_uri_, timestamp_start, timestamp_end);
+
+  MemoryTracker memory_tracker;
+  auto&& [st, array_schema_latest, array_schemas_all, fragment_metadata] =
+      storage_manager_->load_array_schemas_and_fragment_metadata(
+          array_dir, &memory_tracker, enc_key, true);
+  throw_if_not_ok(st);
+
+  tiledb_fragment_max_tile_sizes_t& maxs = *max_tile_sizes;
+  memset(&maxs, 0, sizeof(maxs));
+
+  for (auto& frag : *fragment_metadata) {
+    auto& frags_schema = frag->array_schema();
+
+    if (!frags_schema->dense()) {
+      get_max_tile_size_for_attribute_dim(
+          maxs, *frag, constants::coords, false, false, enc_key);
+    }
+
+    auto& attributes = frags_schema->attributes();
+    for (auto& attrib : attributes) {
```

Comment:
This is confusing as get_max_tile_size_for_attribute_dim both gets the max and mutates the maxs structure... We could make the function return a tiledb_fragment_max_tile_sizes_t object and make a function that merges two object in the tiledb_fragment_max_tile_sizes_t structure. It could be a static max method and the code would now look like:
`current_attr_maxes = get_max_tile_size_for_attribute_dim(...);`
`maxes = tiledb_fragment_max_tile_sizes_t::max(maxes, current_attr_maxes);`

Reply:
method rename should eliminate any confusion

(As far as I can see, adding return of object and additional comparison method/use would require additional unnecessary object instance constructions and data comparisons.)

Reply:
The extra construction would happen on the stack... And we could also live with the extra comparisons if the code is clearer as this isn't a very performance sensitive codepath. We don't need to change it for this review.